Added 3 steps of DP and Fibonacci solution (#260)
cpinitiative/usaco-guide: content/4_Gold/Intro_DP.mdx

</Info>

**Dynamic Programming** (DP) is an important algorithmic technique in Competitive Programming from the gold division to competitions like the International Olympiad of Informatics. By breaking down the full task into sub-problems, DP avoids the redundant computations of brute force solutions.

Although it is not too difficult to grasp the general ideas behind DP, the technique can be used in a diverse range of problems and is a must-know idea for competitors in the USACO Gold division.

> Often, dynamic programming can be split into 3 steps: the base cases, the recurrence, and the answer. 

Reviewer: isn't finding dp state an important part of DP?

## Base Cases

The base cases are the cases of our dynamic programming that cannot be defined in terms of other cases of the same function. Typically, we set these cases manually before performing our recurrence in order to ensure that our recurrence works properly.

## Recurrence/Transition

The recurrence of our dynamic programming is the way we define each case in the terms of other cases of the same function. We always want the recurrence to be made up of cases that we have computed before our current case. 

## The Answer

Lastly, we want to figure out which case of our dynamic programming is the answer to our problem. This is usually an intuitive step of dynamic programming and will almost always come naturally. 

## Solution - The $N$th Fibonacci Term Problem

> A common introductory problem for dynamic programming is finding the N-th Fibonacci term. This can be done with recursion with the formula $F_n$ = $F_{n-1}$ + $F_{n-2}$ in $O(N^2)$ by simply recursing down to 0 for each term. However, notice that we are repeating some steps for each term. In order to reduce our complexity, we can use dynamic programming. 

Reviewer: ```suggestion
A common introductory problem for dynamic programming is finding the N-th Fibonacci term. This can be done with recursion with the formula $F_n = F_{n-1} + F_{n-2}$ in $O(N^2)$ by simply recursing down to 0 for each term. However, notice that we are repeating some steps for each term. In order to reduce our complexity, we can use dynamic programming. 
```

Let $\texttt{dp}[x]$ be the $x$th term in the fibonacci sequence. We can use our recursive formula for our recurrence:

$$
\texttt{dp}[N] = \texttt{dp}[N-1] + \texttt{dp}[N-2]. 
$$

However, notice that for the cases $N = 1$ and $N = 0$, there doesn't exist a $\texttt{dp}[N-2]$. These 2 cases are our base cases. For this problem, $N = 1$ and $N = 0$ are our only base cases, and we can simply set 

> $$
> \texttt{dp}[0] = 0,
> $$
> 
> $$
> \texttt{dp}[1] = 1.
> $$

Reviewer: ```suggestion
$$
\texttt{dp}[0] = 0, \\
\texttt{dp}[1] = 1.
$$
```

</Info>

**Dynamic Programming** (DP) is an important algorithmic technique in Competitive Programming from the gold division to competitions like the International Olympiad of Informatics. By breaking down the full task into sub-problems, DP avoids the redundant computations of brute force solutions.

> Although it is not too difficult to grasp the general ideas behind DP, the technique can be used in a diverse range of problems and is a must-know idea for competitors in the USACO Gold division.

Reviewer 2: These two lines sound kinda clumsy. Maybe change it to something like "Dynamic programming (DP) is an important technique based on the idea of reusing previously computed values. It's used..."

Also, you should use only "DP" or only "dynamic programming" throughout the module for consistency

> Often, dynamic programming can be split into 4 steps: identify the state, the recurrence, the base case, and the answer. 

Reviewer 2: ```suggestion
Often, dynamic programming can be split into 4 steps: identifying the state, finding the recurrence,  finding the base case, and then outputting answer.
```

### State

The state of our dynamic programming is what each spot in our dp array/matrix represents. We want to set up the states such that the answer is one of the states, and there is a clear transition between states.

> One way to think about states is to ask yourself "what information do I need to know to identify my state / progress in solving this problem?" For example, the state for fibonacci could be i, representing that you're trying to find the i'th fibonacci number.

Reviewer 2: Use LaTeX for "i". Also, keep the punctuation and spacing consistent and correct

### Recurrence/Transition

> The recurrence of our dynamic programming is the way we define each case in the terms of other cases of the same function. We always want the recurrence to be made up of cases that we have computed before our current case. The recurrence is largely dependent on what our state is. 

Reviewer 2: I think you should add an example (something like the recurrence for Fibonacci numbers like in the previous section)

### Base Cases

The base cases are the cases of our dynamic programming that cannot be defined in terms of other cases of the same function. Typically, we set these cases manually before performing our recurrence in order to ensure that our recurrence works properly.

### The Answer

Lastly, we want to figure out which case of our dynamic programming is the answer to our problem. This is usually an intuitive step of dynamic programming and will almost always come naturally. 

## Solution - The $N$th Fibonacci Term Problem

A common introductory problem for dynamic programming is finding the N-th Fibonacci term. This can be done with recursion with the formula $F_n = F_{n-1} + F_{n-2}$ in $O(2^n)$ by simply recursing down to 0 for each term. However, notice that we are repeating some steps for each term. In order to reduce our complexity, we can use dynamic programming. 

First, we find the state of our dynamic programming. Here, we will let $\texttt{dp}[x]$ represent the $x$th term in the Fibonacci sequence. 

Next, we need to figure out what our recurrence/transition is. We can use our recursive formula for our recurrence:

$$

> \texttt{dp}[N] = \texttt{dp}[N-1] + \texttt{dp}[N-2]. 

Reviewer 2: You shouldn't have all three of $N$, $n$, and $x$ because it's quite confusing. Maybe use only $N$ and specify $x \leq N$? (Also, you used $i$ previously instead of $x$)

$$

However, notice that for the cases $N = 1$ and $N = 0$, there doesn't exist a $\texttt{dp}[N-2]$. These 2 cases are our base cases. For this problem, $N = 1$ and $N = 0$ are our only base cases, and we can simply set 

$$
\texttt{dp}[0] = 0, \\

> \texttt{dp}[1] = 1.

Reviewer 2: I don't like mixing \texttt and normal LaTeX
